plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java

```diff
@@ -0,0 +1,322 @@
+        ImmutableMap.Builder<ColumnHandle, Domain> primaryKeysDomainBuilder = ImmutableMap.builder();
+        // This value is to build the TupleDomain, but it won't affect the query field in result of the `QueryBuilder#prepareDeleteQuery` and `QueryBuilder#prepareUpdateQuery`
+        Domain dummy = Domain.singleValue(BIGINT, 0L);
+        for (JdbcColumnHandle columnHandle : primaryKeys) {
+            primaryKeysDomainBuilder.put(columnHandle, dummy);
+        }
+        TupleDomain<ColumnHandle> primaryKeysDomain = TupleDomain.withColumnDomains(primaryKeysDomainBuilder.buildOrThrow());
+        JdbcNamedRelationHandle relation = mergeHandle.getTableHandle().getRequiredNamedRelation();
+        this.insertSink = new JdbcPageSink(session, outputHandle, jdbcClient, pageSinkId, queryModifier, JdbcClient::buildInsertSql);
+        this.updateSink = createUpdateSink(session, relation, primaryKeysDomain, primaryKeys, jdbcClient, pageSinkId, queryModifier, queryBuilder);
+        this.deleteSink = createDeleteSink(session, relation, primaryKeysDomain, primaryKeys, jdbcClient, pageSinkId, queryModifier, queryBuilder);
+    }
+
+    private static int[] getUpdateAssignmentChannels(JdbcOutputTableHandle outputTableHandle, List<JdbcColumnHandle> primaryKeys)
+    {
+        Set<String> primaryKeyNames = primaryKeys.stream()
+                .map(JdbcColumnHandle::getColumnName)
+                .collect(toImmutableSet());
+        List<String> allDataColumns = outputTableHandle.getColumnNames();
+
+        ImmutableList.Builder<Integer> updateAssignmentChannelBuilder = ImmutableList.builder();
+        for (int channel = 0; channel < allDataColumns.size(); channel++) {
+            String column = allDataColumns.get(channel);
+            if (!primaryKeyNames.contains(column)) {
+                updateAssignmentChannelBuilder.add(channel);
+            }
+        }
+        return updateAssignmentChannelBuilder.build().stream()
+                .mapToInt(Integer::intValue)
+                .toArray();
+    }
+
+    private static ConnectorPageSink createUpdateSink(
+            ConnectorSession session,
+            JdbcNamedRelationHandle relation,
+            TupleDomain<ColumnHandle> domain,
+            List<JdbcColumnHandle> primaryKeys,
+            JdbcClient jdbcClient,
+            ConnectorPageSinkId pageSinkId,
+            RemoteQueryModifier remoteQueryModifier,
+            QueryBuilder queryBuilder)
+    {
+        List<JdbcColumnHandle> columns = jdbcClient.getColumns(session, new JdbcTableHandle(relation.getSchemaTableName(), relation.getRemoteTableName(), Optional.empty()));
+
+        ImmutableList.Builder<JdbcAssignmentItem> assignmentItemBuilder = ImmutableList.builder();
+        ImmutableList.Builder<String> columnNamesBuilder = ImmutableList.builder();
+        ImmutableList.Builder<Type> columnTypesBuilder = ImmutableList.builder();
+        QueryParameter dummy = new QueryParameter(BIGINT, Optional.empty());
```

Comment:
This requires a comment like above with tuple domain. Let's write that value won't be used as the real value will be passed to `storeMergedRows`